vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,799 @@
+                                    | None -> displayContext
+                                 
+                                let typeLayout = ty.FormatLayout(displayContext)
+                                let taggedText = ResizeArray()
+                                Layout.renderL (Layout.taggedTextListR taggedText.Add) typeLayout |> ignore
+                                let navigation = QuickInfoNavigation(gotoDefinitionService, document, func.DeclarationLocation)
+                                // Because the data is available notify that this line should be updated, displaying the results
+                                return Some (taggedText, navigation)
+                            | None -> 
+                                logWarningf "Couldn't acquire CodeLens data for function %A" func
+                                return None
+                        else return None
+                    with e -> 
+                        logErrorf "Error in lazy code lens computation. %A" e
+                        return None
+                }
+
+            for symbolUse in symbolUses do
+                if symbolUse.IsFromDefinition then
+                    match symbolUse.Symbol with
+                    | :? FSharpEntity as entity ->
+                        for func in entity.MembersFunctionsAndValues do
+                            // Regardles of whether we are in a async maybe, we don't want to abort the whole process due to a single empty option.
+                            let rawElement = visit func.DeclarationLocation.Start parsedInput
+                            match rawElement with
+                            | None -> ()
+                            | Some declarationRange ->
+                                let! declarationSpan = FSharpRangeToSpan textSnapshot declarationRange
+                                let funcID = func.FullName
+                                let fullDeclarationText = (textSnapshot.GetText declarationSpan).Replace(func.CompiledName, funcID)
+                                let fullTypeSignature = func.FullType.ToString()
+                                // Try to re-use the last results
+                                if lastResults.ContainsKey(fullDeclarationText) then
+                                    // Make sure that the results are usable
+                                    let lastTrackingSpan, codeLens = lastResults.[fullDeclarationText]
+                                    if codeLens.FullTypeSignature = fullTypeSignature then
+                                        // The results can be reused because the signature is the same
+                                        if codeLens.Computed then
+                                            // Just re-use the old results, changed nothing
+                                            newResults.[fullDeclarationText] <- (lastTrackingSpan, codeLens)
```

Comment:
This nesting is a big crazy - how many levels deep is this?

Reply:
Welcome to the crazy caching algorithm of my Code Lens ;)